### PR TITLE
[REVIEW] Adding support for unsigned int

### PR DESCRIPTION
**Pull request by rgsl888prabhu**
Adds support of `unsigned int` to cudf, updated validations and test cases to accommodate it.

And instead of `-1` as sentinel value, `max(int64)` is being used in most of the cases.

### CODE REVIEW

File: python/cudf/cudf/tests/test_binops.py [hunk 1]
```diff
@@ -88,7 +88,16 @@ def test_series_binop_scalar(nelem, binop, obj_class):
 _bitwise_binops = [operator.and_, operator.or_, operator.xor]
 
 
-_int_types = ["int8", "int16", "int32", "int64", "longlong"]
+_int_types = [
```
Comment (shwina) [inline python/cudf/cudf/tests/test_binops.py:91]: So, in general, it looks like a lot of tests need a list of all numeric types (or just signed/unsigned types). Instead of repeating everywhere, let's have a `tests.utils.SIGNED_TYPES`, `tests.utils.UNSIGNED_TYPES`, `tests.utils.NUMERIC_TYPES`, etc.?

Comment (shwina) [reply python/cudf/cudf/tests/test_binops.py:91]: Or maybe `cudf.utils.dtypes.*TYPES`.

Comment (rgsl888prabhu) [reply python/cudf/cudf/tests/test_binops.py:91]: That works, it would be better.

File: python/cudf/cudf/utils/dtypes.py [hunk 2]
```diff
@@ -228,7 +232,20 @@ def min_signed_type(x, min_size=8):
...
+def min_unsigned_type(x, min_size=8):
+    """
+    Return the smallest *unsigned* integer dtype
+    that can represent the integer ``x``
+    """
+    for int_dtype in np.sctypes["uint"]:
+        if (np.dtype(int_dtype).itemsize * 8) >= min_size:
+            if np.iinfo(int_dtype).min <= x <= np.iinfo(int_dtype).max:
+                return int_dtype
+    # resort to using `int64` and let numpy raise appropriate exception:
```
Comment (shwina) [inline python/cudf/cudf/utils/dtypes.py:244]:
```suggestion
    # resort to using `uint64` and let numpy raise appropriate exception:
```

File: cpp/include/cudf/utilities/traits.hpp [hunk 3]
```diff
@@ -150,6 +171,20 @@ constexpr inline bool is_numeric(data_type type)
...
+/**
+ * @brief Indicates whether `type` is a unsigned numeric `data_type`.
+ *
+ * "Unsigned Numeric" types are fundamental integral types such as `UINT*`.
+ *
+ * @param type The `data_type` to verify
+ * @return true `type` is unsigned numeric
+ * @return false `type` is signed numeric
+ **/
+constexpr inline bool is_unsigned_numeric(data_type type)
```
Comment (jrhemstad) [inline cpp/include/cudf/utilities/traits.hpp:183]: Can we just call this `is_unsigned`?

File: python/cudf/cudf/tests/test_categorical.py [hunk 4]
```diff
@@ -553,6 +553,7 @@ def test_categorical_set_categories_categoricals(data, new_categories):
         pd.Series(["a", "1", "22", "1", "aa"]),
         pd.Series(["a", "1", "22", "1", "aa"], dtype="category"),
         pd.Series([1, 2, 3, 4], dtype="int64"),
+        pd.Series([1, 2, 3, 4], dtype="uint64"),
```
Comment (cwharris) [inline python/cudf/cudf/tests/test_categorical.py:556]: Should some of these be negative? Doesn't seem like a valid use case for categorical, but maybe worth mentioning.

File: python/cudf/cudf/tests/test_binops.py (hunk 1 above)
Comment (cwharris) [reply python/cudf/cudf/tests/test_binops.py:91]: we could get clever do something like `cudf.utils.dtypes.ALL_TYPES - ("str", float64")`. :imp:

File: python/cudf/cudf/tests/test_index.py [hunk 5]
```diff
@@ -141,6 +141,7 @@ def test_categorical_index():
 def test_pandas_as_index():
     # Define Pandas Indexes
     pdf_int_index = pd.Int64Index([1, 2, 3, 4, 5])
+    pdf_uint_index = pd.UInt64Index([1, 2, 3, 4, 5], dtype="uint64")
```
Comment (cwharris) [inline python/cudf/cudf/tests/test_index.py:144]: Hoe come `UInt64Index` needs a dtype param? `uint64` is in the name of the class... ? Maybe we could default `UInt64Index` with a dtype of `uint64`?

File: python/cudf/cudf/tests/test_repr.py [hunk 6]
```diff
@@ -27,25 +27,36 @@
 def test_null_series(nrows, dtype):
     size = 5
     mask = utils.random_bitmask(size)
-    data = cudf.Series(np.random.randint(0, 128, size))
+    data = cudf.Series(np.random.randint(1, 9, size))
```
Comment (cwharris) [inline python/cudf/cudf/tests/test_repr.py:30]: Was this change intentional? I'm not sure I understand it if it is.

File: python/cudf/cudf/tests/test_repr.py [hunk 7]
```diff
@@ -56,7 +67,7 @@ def test_null_dataframe(ncols):
     pdfrepr = pdf.__repr__()
     pdfrepr = pdfrepr.replace("NaN", "null")
     pdfrepr = pdfrepr.replace("NaT", "null")
-    pdfrepr = pdfrepr.replace("-1", "null")
+    # pdfrepr = pdfrepr.replace("-1", "null")
```
Comment (cwharris) [inline python/cudf/cudf/tests/test_repr.py:70]: Why not remove the line altogether? We'll still have it in git history.

File: python/cudf/cudf/tests/test_sorting.py [hunk 8]
```diff
@@ -57,7 +64,7 @@ def test_series_argsort(nelem, dtype, asc):
     if asc:
         expected = np.argsort(sr.to_array(), kind="mergesort")
     else:
-        expected = np.argsort(-sr.to_array(), kind="mergesort")
+        expected = np.argsort(sr.to_array() * -1, kind="mergesort")
```
Comment (cwharris) [inline python/cudf/cudf/tests/test_sorting.py:67]: Thanks for making this more obvious :grin:  :pray:

**Comment by harrism** [comment]
Will this close #2819?

File: python/cudf/cudf/tests/test_dataframe.py [hunk 9]
```diff
@@ -1791,9 +1724,7 @@ def test_series_hash_encode(nrows):
     assert enc_with_name_arr[0] != enc_arr[0]
 
 
-@pytest.mark.parametrize(
-    "dtype", ["int8", "int16", "int32", "int64", "float32", "float64"]
-)
+@pytest.mark.parametrize("dtype", NUMERIC_TYPES + ["bool"])
```
Comment (kkraus14) [inline python/cudf/cudf/tests/test_dataframe.py:1727]:
```suggestion
@pytest.mark.parametrize("dtype", NUMERIC_TYPES + {"bool"})
```

File: python/cudf/cudf/tests/test_dataframe.py [hunk 10]
```diff
@@ -1981,9 +1897,7 @@ def test_arrow_handle_no_index_name(pdf, gdf):
 @pytest.mark.parametrize("num_rows", [1, 3, 10, 100])
 @pytest.mark.parametrize("num_bins", [1, 2, 4, 20])
 @pytest.mark.parametrize("right", [True, False])
-@pytest.mark.parametrize(
-    "dtype", ["int8", "int16", "int32", "int64", "float32", "float64"]
-)
+@pytest.mark.parametrize("dtype", NUMERIC_TYPES + ["bool"])
```
Comment (kkraus14) [inline python/cudf/cudf/tests/test_dataframe.py:1900]:
```suggestion
@pytest.mark.parametrize("dtype", NUMERIC_TYPES + {"bool"})
```

File: python/cudf/cudf/tests/test_dataframe.py [hunk 11]
```diff
@@ -2453,23 +2352,7 @@ def test_dataframe_empty_sort_index():
...
-        "float32",
-        "float64",
-        "category",
-        "datetime64[s]",
-        "datetime64[ms]",
-        "datetime64[us]",
-        "datetime64[ns]",
-    ],
-)
+@pytest.mark.parametrize("dtype", dtypes + ["category"])
```
Comment (kkraus14) [inline python/cudf/cudf/tests/test_dataframe.py:2355]:
```suggestion
@pytest.mark.parametrize("dtype", dtypes + {"category"})
```

File: python/cudf/cudf/tests/test_dataframe.py [hunk 12]
```diff
@@ -4192,28 +4066,15 @@ def test_tolist_mixed_nulls():
...
-        "datetime64[ns]",
-    ],
-)
+@pytest.mark.parametrize("dtype", NUMERIC_TYPES)
+@pytest.mark.parametrize("as_dtype", ALL_TYPES)
 def test_df_astype_numeric_to_all(dtype, as_dtype):
-    if "int" in dtype:
+    if "uint" not in dtype and "uint" in as_dtype:
+        # can't convert int to unisgned
+        return
```
Comment (kkraus14) [inline python/cudf/cudf/tests/test_dataframe.py:4074]: Why can't we convert `int` to `uint`?

File: python/cudf/cudf/tests/test_dataframe.py [hunk 13]
```diff
@@ -4369,6 +4230,10 @@ def test_df_astype_to_categorical_ordered(ordered):
         ("int16", {}),
```
Comment (kkraus14) [inline python/cudf/cudf/tests/test_dataframe.py:4230]: May want to programmatically define by looping through `ALL_TYPES` and then special casing `category`.

File: python/cudf/cudf/tests/test_feather.py [hunk 14]
```diff
@@ -24,7 +24,7 @@
 
 @pytest.fixture(params=[0, 1, 10, 100])
 def pdf(request):
-    types = ["bool", "int8", "int16", "int32", "int64", "float32", "float64"]
+    types = NUMERIC_TYPES + ["bool"]
```
Comment (kkraus14) [inline python/cudf/cudf/tests/test_feather.py:27]:
```suggestion
    types = NUMERIC_TYPES | {"bool"}
```

File: python/cudf/cudf/tests/test_hdf.py [hunk 15]
```diff
@@ -21,16 +21,7 @@
...
-        "bool",
-        "int8",
-        "int16",
-        "int32",
-        "int64",
-        "float32",
-        "float64",
-        "datetime64[ms]",
-    ]
+    types = NUMERIC_TYPES + DATETIME_TYPES + ["bool"]
```
Comment (kkraus14) [inline python/cudf/cudf/tests/test_hdf.py:24]:
```suggestion
    types = NUMERIC_TYPES | DATETIME_TYPES | {"bool"}
```

File: python/cudf/cudf/tests/test_hdfs.py [hunk 16]
```diff
@@ -38,7 +38,7 @@ def pdf(scope="module"):
     df = pd.DataFrame()
     df["Integer"] = np.array([2345, 11987, 9027, 9027])
     df["Float"] = np.array([9.001, 8.343, 6, 2.781])
-    df["Integer2"] = np.array([2345, 106, 2088, 789277])
+    df["Integer2"] = np.array([2345, 106, 2088, 789277]).astype("uint64")
```
Comment (kkraus14) [inline python/cudf/cudf/tests/test_hdfs.py:41]:
```suggestion
    df["Integer2"] = np.array([2345, 106, 2088, 789277], dtype="uint64")
```

File: python/cudf/cudf/tests/test_index.py [hunk 17]
```diff
@@ -491,19 +495,7 @@ def test_index_where(data, condition, other, error):
...
-        "int16",
-        "int32",
-        "int64",
-        "float32",
-        "float64",
-        "str",
-        "category",
-    ],
-)
+@pytest.mark.parametrize("dtype", NUMERIC_TYPES + OTHER_TYPES)
```
Comment (kkraus14) [inline python/cudf/cudf/tests/test_index.py:498]:
```suggestion
@pytest.mark.parametrize("dtype", NUMERIC_TYPES | OTHER_TYPES)
```

File: python/cudf/cudf/tests/test_joining.py [hunk 18]
```diff
@@ -1097,13 +1094,17 @@ def test_typecast_on_join_no_float_round():
...
+        (np.dtype("uint32"), np.dtype("uint64")),
         (np.dtype("float32"), np.dtype("float64")),
         (np.dtype("int32"), np.dtype("float32")),
+        (np.dtype("uint32"), np.dtype("float32")),
     ],
 )
 def test_typecast_on_join_overflow_unsafe(dtypes):
     dtype_l, dtype_r = dtypes
-    if dtype_l.kind == "i":
+    if dtype_l.kind in "iu":
```
Comment (kkraus14) [inline python/cudf/cudf/tests/test_joining.py:1107]:
```suggestion
    if dtype_l.kind in {"i", "u"}:
```

File: python/cudf/cudf/tests/test_json.py [hunk 19]
```diff
@@ -23,16 +23,7 @@ def make_numeric_dataframe(nrows, dtype):
...
-        "bool",
-        "int8",
-        "int16",
-        "int32",
-        "int64",
-        "float32",
-        "float64",
-        "datetime64[ms]",
-    ]
+    types = NUMERIC_TYPES + DATETIME_TYPES + ["bool"]
```
Comment (kkraus14) [inline python/cudf/cudf/tests/test_json.py:26]:
```suggestion
    types = NUMERIC_TYPES | DATETIME_TYPES | {"bool"}
```

File: python/cudf/cudf/tests/test_reductions.py [hunk 20]
```diff
@@ -19,9 +19,11 @@
 
 @pytest.mark.parametrize("dtype,nelem", params)
 def test_sum(dtype, nelem):
+    dtype = np.dtype(dtype).type
     data = gen_rand(dtype, nelem)
     sr = Series(data)
 
+    # got = dtype(sr.sum())
```
Comment (kkraus14) [inline python/cudf/cudf/tests/test_reductions.py:26]:
```suggestion
```

File: python/cudf/cudf/tests/test_reductions.py [hunk 21]
```diff
@@ -47,7 +49,8 @@ def test_sum_string():
 
 @pytest.mark.parametrize("dtype,nelem", params)
 def test_product(dtype, nelem):
-    if np.dtype(dtype).kind == "i":
+    dtype = np.dtype(dtype).type
+    if np.dtype(dtype).kind in "ui":
```
Comment (kkraus14) [inline python/cudf/cudf/tests/test_reductions.py:53]:
```suggestion
    if np.dtype(dtype).kind in {"u", "i"}:
```

File: python/cudf/cudf/tests/test_reductions.py [hunk 22]
```diff
@@ -69,13 +72,15 @@ def test_product(dtype, nelem):
...
+    dtype = np.dtype(dtype).type
     data = gen_rand(dtype, nelem)
     sr = Series(data)
 
     got = sr.sum_of_squares()
+    # got = dtype(got)
     expect = (data ** 2).sum()
 
-    if np.dtype(dtype).kind == "i":
+    if np.dtype(dtype).kind in "ui":
```
Comment (kkraus14) [inline python/cudf/cudf/tests/test_reductions.py:83]:
```suggestion
    if np.dtype(dtype).kind in {"u", "i"}:
```

File: python/cudf/cudf/tests/test_replace.py [hunk 23]
```diff
@@ -299,7 +295,10 @@ def test_fillna_string(fill_type, inplace):
     assert_eq(expect, got)
 
 
-@pytest.mark.parametrize("data_dtype", ["int8", "int16", "int32", "int64"])
+@pytest.mark.parametrize(
+    "data_dtype",
+    ["int8", "int16", "int32", "int64", "uint8", "uint16", "uint32", "uint64"],
```
Comment (kkraus14) [inline python/cudf/cudf/tests/test_replace.py:300]: Use INTEGER_TYPES

File: python/cudf/cudf/tests/test_replace.py [hunk 24]
```diff
@@ -491,16 +486,16 @@ def test_numeric_series_replace_dtype(series_dtype, replacement):
         sr.replace([0, 1], [replacement])
 
     # Both lists of equal length
-    if (np.dtype(type(replacement)).kind == "f" and sr.dtype.kind == "i") or (
+    if (np.dtype(type(replacement)).kind == "f" and sr.dtype.kind in "iu") or (
```
Comment (kkraus14) [inline python/cudf/cudf/tests/test_replace.py:489]:
```suggestion
    if (np.dtype(type(replacement)).kind == "f" and sr.dtype.kind in {"i", "u"}) or (
```

File: python/cudf/cudf/tests/test_repr.py [hunk 25]
```diff
@@ -9,43 +9,48 @@
...
-    "int16",
-    "int32",
-    "int64",
-    "float32",
-    "float64",
-    "datetime64[ns]",
-    "str",
-    "category",
-]
+repr_categories = utils.NUMERIC_TYPES + ["str", "category", "datetime64[ns]"]
```
Comment (kkraus14) [inline python/cudf/cudf/tests/test_repr.py:12]:
```suggestion
repr_categories = utils.NUMERIC_TYPES | {"str", "category", "datetime64[ns]"}
```

File: python/cudf/cudf/tests/test_reshape.py [hunk 26]
```diff
@@ -5,24 +5,18 @@
...
-        "int8",
-        "int16",
-        "int32",
-        "int64",
-        "float32",
-        "float64",
-        "datetime64[ms]",
-    ],
-)
+@pytest.mark.parametrize("dtype", NUMERIC_TYPES + DATETIME_TYPES)
```
Comment (kkraus14) [inline python/cudf/cudf/tests/test_reshape.py:19]:
```suggestion
@pytest.mark.parametrize("dtype", NUMERIC_TYPES | DATETIME_TYPES)
```

File: python/cudf/cudf/tests/test_reshape.py [hunk 27]
```diff
@@ -76,16 +70,9 @@ def test_melt(nulls, num_id_vars, num_value_vars, num_rows, dtype):
...
-        "int32",
-        "int64",
-        "float32",
-        "float64",
-        "datetime64[ms]",
-        pytest.param("str", marks=pytest.mark.xfail()),
-    ],
+    NUMERIC_TYPES
+    + DATETIME_TYPES
+    + [pytest.param("str", marks=pytest.mark.xfail())],
```
Comment (kkraus14) [inline python/cudf/cudf/tests/test_reshape.py:75]:
```suggestion
    NUMERIC_TYPES
    | DATETIME_TYPES
    | {pytest.param("str", marks=pytest.mark.xfail())},
```

File: python/cudf/cudf/tests/test_reshape.py [hunk 28]
```diff
@@ -120,17 +107,7 @@ def test_df_stack(nulls, num_cols, num_rows, dtype):
...
-        "int8",
-        "int16",
-        "int32",
-        "int64",
-        "float32",
-        "float64",
-        "datetime64[ms]",
-        "category",
-    ],
+    "dtype", NUMERIC_TYPES + DATETIME_TYPES + ["category"]
```
Comment (kkraus14) [inline python/cudf/cudf/tests/test_reshape.py:110]:
```suggestion
    "dtype", NUMERIC_TYPES | DATETIME_TYPES | {"category"}
```

File: python/cudf/cudf/tests/test_scalar.py [hunk 29]
```diff
@@ -27,6 +28,21 @@
         np.int64(42),
         np.iinfo(np.int64).min,
         np.iinfo(np.int64).max,
+        np.uint8(0),
+        np.uint8(1),
+        np.uint8(-1),
```
Comment (kkraus14) [inline python/cudf/cudf/tests/test_scalar.py:33]:
```suggestion
        np.uint8(255),
```

File: python/cudf/cudf/tests/test_scalar.py [hunk 30]
```diff
@@ -27,6 +28,21 @@
...
+        np.iinfo(np.uint8).min,
+        np.iinfo(np.uint8).max,
+        np.uint16(1),
+        np.iinfo(np.uint16).min,
+        np.iinfo(np.uint16).max,
+        np.uint32(42),
+        np.uint32(-42),
```
Comment (kkraus14) [inline python/cudf/cudf/tests/test_scalar.py:40]:
```suggestion
        np.uint32(4294967254),
```

File: python/cudf/cudf/tests/test_scalar.py [hunk 31]
```diff
@@ -27,6 +28,21 @@
...
+        np.iinfo(np.uint32).min,
+        np.iinfo(np.uint32).max,
+        np.uint64(42),
+        np.iinfo(np.uint64).min,
+        np.dtype("uint64").type(np.iinfo(np.uint64).max),
```
Comment (kkraus14) [inline python/cudf/cudf/tests/test_scalar.py:45]:
```suggestion
        np.uint64(np.iinfo(np.uint64).max),
```

File: python/cudf/cudf/tests/test_scalar.py [hunk 32]
```diff
@@ -72,22 +88,7 @@ def test_round_trip_scalar(value):
...
-        "float32",
-        "float64",
-        "datetime64[s]",
-        "datetime64[ms]",
-        "datetime64[us]",
-        "datetime64[ns]",
-        "object",
-    ],
-)
+@pytest.mark.parametrize("dtype", NUMERIC_TYPES + DATETIME_TYPES + ["object"])
```
Comment (kkraus14) [inline python/cudf/cudf/tests/test_scalar.py:91]:
```suggestion
@pytest.mark.parametrize("dtype", NUMERIC_TYPES | DATETIME_TYPES | {"object"})
```

File: python/cudf/cudf/tests/test_sorting.py [hunk 33]
```diff
@@ -169,18 +176,7 @@ def test_dataframe_nsmallest_sliced(counts, sliceobj):
...
-        "int8",
-        "int16",
-        "int32",
-        "int64",
-        "float32",
-        "float64",
-        "datetime64[ms]",
-    ],
-)
+@pytest.mark.parametrize("dtype", NUMERIC_TYPES + DATETIME_TYPES)
```
Comment (kkraus14) [inline python/cudf/cudf/tests/test_sorting.py:179]:
```suggestion
@pytest.mark.parametrize("dtype", NUMERIC_TYPES | DATETIME_TYPES)
```

File: python/cudf/cudf/tests/test_string.py [hunk 34]
```diff
@@ -160,22 +160,14 @@ def test_string_repr(ps_gs, item):
...
-        "int8",
-        "int16",
-        "int32",
-        "int64",
-        "float32",
-        "float64",
-        "bool",
-        "datetime64[ms]",
-    ],
+    "dtype", NUMERIC_TYPES + DATETIME_TYPES + ["bool", "object", "str"]
```
Comment (kkraus14) [inline python/cudf/cudf/tests/test_string.py:163]:
```suggestion
    "dtype", NUMERIC_TYPES | DATETIME_TYPES | {"bool", "object", "str"}
```

File: python/cudf/cudf/tests/test_string.py [hunk 35]
```diff
@@ -205,19 +197,7 @@ def test_string_astype(dtype):
...
-        "object",
-        "int16",
-        "int32",
-        "int64",
-        "float32",
-        "float64",
-        "bool",
-        "datetime64[ms]",
-    ],
+    "dtype", NUMERIC_TYPES + DATETIME_TYPES + ["bool", "object", "str"]
```
Comment (kkraus14) [inline python/cudf/cudf/tests/test_string.py:200]:
```suggestion
    "dtype", NUMERIC_TYPES | DATETIME_TYPES | {"bool", "object", "str"}
```

File: python/cudf/cudf/tests/test_string.py [hunk 36]
```diff
@@ -230,23 +210,15 @@ def test_string_empty_astype(dtype):
...
-        "int16",
-        "int32",
-        "int64",
-        "float32",
-        "float64",
-        "bool",
-        "datetime64[ms]",
-    ],
-)
+@pytest.mark.parametrize("dtype", NUMERIC_TYPES + DATETIME_TYPES + ["bool"])
```
Comment (kkraus14) [inline python/cudf/cudf/tests/test_string.py:213]:
```suggestion
@pytest.mark.parametrize("dtype", NUMERIC_TYPES | DATETIME_TYPES | {"bool"})
```

File: python/cudf/cudf/tests/test_string.py [hunk 37]
```diff
@@ -271,19 +243,7 @@ def test_string_numeric_astype(dtype):
...
```
Comment (kkraus14) [inline python/cudf/cudf/tests/test_string.py:246]:
```suggestion
@pytest.mark.parametrize("dtype", NUMERIC_TYPES | DATETIME_TYPES | {"bool"})
```

**Comment by rgsl888prabhu** [comment]
> Will this close #2819?

It will, but there are other issues such as #5352 and #5351 which still requires changes in libcudf side.

**Comment by rgsl888prabhu** [comment]
> I'm super uncomfortable with using 0 as a sentinel for nulls throughout the tests. It seems very error prone and fragile. Any ideas on what we can do instead? If we wait for the Pandas 1.0 PR do the new Pandas nullable types help?

 It might help, but that would require lot of additional changes to testing to accommodate that change and how close are we with supporting pandas 1.0?

File: python/cudf/cudf/_lib/reduce.pyx [hunk 38]
```diff
@@ -31,7 +31,7 @@ def reduce(reduction_op, Column incol, dtype=None, **kwargs):
 
     col_dtype = incol.dtype
     if reduction_op in ['sum', 'sum_of_squares', 'product']:
-        col_dtype = np.find_common_type([col_dtype], [np.int64])
+        col_dtype = np.find_common_type([col_dtype], [np.uint64])
```
[the remark this answers is not in the document]
Comment (rgsl888prabhu) [reply python/cudf/cudf/_lib/reduce.pyx:34]: Earlier with np.int64, we would end up getting signed int instead of unsigned
```
>>> np.find_common_type([np.uint32], [np.int64])
dtype('int64')
```
but with unit64, it will return unsigned for unsigned types.
```
>>> np.find_common_type([np.int64], [np.uint64])
dtype('int64')
>>> np.find_common_type([np.uint32], [np.uint64])
dtype('uint32')

```

**Comment by shwina** [comment]
> I'm super uncomfortable with using 0 as a sentinel for nulls throughout the tests. It seems very error prone and fragile. Any ideas on what we can do instead?

I agree 100% here. Here are some additional considerations:

1. `0` should be considered as good a sentinel value as any other, i.e., just changing our sentinel value to something else doesn't guarantee our tests aren't broken.

2. Regarding Pandas 1.0, do they support nulls for all the operations that we have? If so, do their null semantics match ours? If not, this will be difficult to incorporate in the short term.

I think that our feature set is at a point where comparing with Pandas isn't sufficient and/or possible in many cases. For these cases, we should determine what we think is the right behaviour and test against _that_, rather than what Pandas does for some different (but closely related) case.

That being said, I'm definitely not advocating for testing against Pandas. Where possible, we should definitely still do that.

File: python/cudf/cudf/utils/dtypes.py [hunk 39]
```diff
@@ -228,7 +232,20 @@ def min_signed_type(x, min_size=8):
...
+    """
+    for int_dtype in np.sctypes["uint"]:
+        if (np.dtype(int_dtype).itemsize * 8) >= min_size:
+            if np.iinfo(int_dtype).min <= x <= np.iinfo(int_dtype).max:
+                return int_dtype
+    # resort to using `uint64` and let numpy raise appropriate exception:
+    return np.uint64(x).dtype
+
+
+def min_numeric_column_type(x, expected_type):
```
[the remark this answers is not in the document]
Comment (rgsl888prabhu) [reply python/cudf/cudf/utils/dtypes.py:248]: https://github.com/rapidsai/cudf/blob/94c9eb1c9061b1f1195ebd5d31e5167dcec1eb02/python/cudf/cudf/core/column/numerical.py#L481

File: python/cudf/cudf/core/column/numerical.py [hunk 40]
```diff
@@ -107,7 +111,7 @@ def normalize_binop_value(self, other):
             if other_dtype == np.dtype("float16"):
                 other = np.dtype("float32").type(other)
                 other_dtype = other.dtype
-            if other_dtype.kind in "u":
+            if self.dtype.kind in "b":
```
[the remark this answers is not in the document]
Comment (rgsl888prabhu) [reply python/cudf/cudf/core/column/numerical.py:114]: To be in sync with pandas, as pandas returns signed output. If we don't changes the type, it will return unsigned types.

https://github.com/rapidsai/cudf/blob/d73f7ae492dfaf691b08dcca566a48c8a21ad4f0/python/cudf/cudf/tests/test_binops.py#L659

File: python/cudf/cudf/tests/test_binops.py [hunk 41]
```diff
@@ -88,7 +88,16 @@ def test_series_binop_scalar(nelem, binop, obj_class):
...
+    "int8",
+    "int16",
+    "int32",
+    "int64",
+    "longlong",
+    "uint8",
+    "uint16",
+    "uint32",
```
[the remark this answers is not in the document]
Comment (rgsl888prabhu) [reply python/cudf/cudf/tests/test_binops.py:99]: I thought `longlong` was kept intentionally, will remove from list.

[the remark this answers is not in the document]
Comment (rgsl888prabhu) [reply python/cudf/cudf/tests/test_binops.py:99]: numpy doesn't support uint64 for this operation
```
>>> import numpy as np
>>> import operator as op
>>> op.and_
<built-in function and_>
>>> a = np.array([1], dtype='int8')
>>> b = np.array([1], dtype="uint64")
>>> op.and_(a, b)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: ufunc 'bitwise_and' not supported for the input types, and the inputs could not be safely coerced to any supported types according to the casting rule ''safe''
```

so thought it shouldn't be added.

File: python/cudf/cudf/tests/test_udf_binops.py [hunk 42]
```diff
@@ -18,7 +18,17 @@
...
+    "int16",
+    "int32",
+    "int64",
+    "uint8",
+    "uint16",
+    "uint32",
+    "uint64",
+    "float32",
+    "float64",
+]
```
[the remark this answers is not in the document]
Comment (rgsl888prabhu) [reply python/cudf/cudf/tests/test_udf_binops.py:31]: uint8 is there, but int8 is not supported in ptx https://github.com/rapidsai/cudf/blob/ad9dea1235d72674f99c6d626a45d4e1296d362d/cpp/src/binaryop/binaryop.cpp#L354

**Comment by rgsl888prabhu** [comment]
@kkraus14 @shwina As a temporary solution till we get Pandas 1.0 support, shall we use max of the col dtype to signify `null`. At least this value will not be as common as `0`.

**Comment by kkraus14** [comment]
> @kkraus14 @shwina As a temporary solution till we get Pandas 1.0 support, shall we use max of the col dtype to signify `null`. At least this value will not be as common as `0`.

It would be slightly better, but it still feels extremely fragile

**Comment by shwina** [comment]
> @kkraus14 @shwina As a temporary solution till we get Pandas 1.0 support, shall we use max of the col dtype to signify null. At least this value will not be as common as 0.

Practically speaking, yes, I think this is better than 0, but theoretically, `max(int64)` isn't any better than `0`. Neither is `17` or anything else for that matter :)

Suppose we decided to re-write the tests to use Pandas nullable types where possible, and our own test data otherwise, what's the level of effort involved? How many tests would this touch? Is this something that a couple of us could pool effort on and pull off?

**Comment by rgsl888prabhu** [comment]
Just to test, I just changed sentinel value to something random, and I can observe around 10 - 15 tests failing, there can be more than that. But it feels like doable. As @brandon-b-miller mentioned in #5388 , we might have to introduce dtypes to handle nulls, which might add additional burden on modifying these test cases.

**Comment by kkraus14** [comment]
Lets use `max(uint64)` as the sentinel value for now (where we can) for the tests, and we'll file issues and follow up in using Pandas nullable types / rolling our own null behaviors in tests as needed to move away from the sentinel value based testing.

**Comment by rgsl888prabhu** [comment]
Changed the sentinel value to be `max(int64)` rather than `max(uint64)` as the value of an int can't be greater than `sys.maxsize`, else any conversion will fail with following error.
```
>>> np.int8(np.iinfo(np.uint64).max)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
OverflowError: Python int too large to convert to C long
```

I have used this sentinel value in most of the places apart from places were float column is being created and then it is casted to integer.

File: python/cudf/cudf/core/column/numerical.py [hunk 43]
```diff
@@ -254,7 +254,7 @@ def default_na_value(self):
         if dkind == "f":
             return self.dtype.type(np.nan)
         elif dkind in {"u", "i"}:
-            return self.dtype.type(0)
+            return self.dtype.type(np.iinfo("int64").max)
```
Comment (kkraus14) [inline python/cudf/cudf/core/column/numerical.py]: This doesn't seem safe. What if the dtype is int8 or int16 or uint32?

File: python/cudf/cudf/utils/cudautils.py [hunk 44]
```diff
@@ -76,25 +76,7 @@ def expand_mask_bits(size, bits):
...
-    if N > 0:
-        if i < in_col.size:
-            out_col[i] = in_col[i - N]
-        if i < N:
-            out_col[i] = 0
-    else:
-        if i <= (in_col.size + N):
-            out_col[i] = in_col[i - N]
-        if i >= (in_col.size + N) and i < in_col.size:
-            out_col[i] = 0
```
Comment (kkraus14) [inline python/cudf/cudf/utils/cudautils.py]: Was this function no longer used?

Comment (rgsl888prabhu) [reply python/cudf/cudf/utils/cudautils.py]: It was a stale code which wasn't being used by any APIs.

File: python/cudf/cudf/core/column/numerical.py (hunk 43 above)
Comment (rgsl888prabhu) [reply python/cudf/cudf/core/column/numerical.py]: It will appropriately change to `-1` or `255` in case of `int8` or `uint8`.

Comment (kkraus14) [reply python/cudf/cudf/core/column/numerical.py]: I think we should just more explicitly use `min(dtype)` for signed integers and `max(dtype)` for unsigned integers. Using `max(int64)` for `uint64` feels gross

Comment (rgsl888prabhu) [reply python/cudf/cudf/core/column/numerical.py]: Changed the sentinel value to be max(int64) rather than max(uint64) as the value of an int can't be greater than sys.maxsize, atleast in case `uint64`.

Comment (rgsl888prabhu) [reply python/cudf/cudf/core/column/numerical.py]: That might work, will take a look.

**Comment by kkraus14** [comment]
@shwina can you review again before we merge?

**Comment by kkraus14** [comment]
> @shwina can you review again before we merge?

connected offline and confirmed this is good to go so merging!
